### PR TITLE
crimson/tools: stress tool update

### PR DESCRIPTION
**Pull request by aisakaki**
Support perf analyse.

Support rand read base case.

Support different types of base case threads running parallel and
users can set the running clients ratio of different types of cases
such as 75% of read and 25% of write.

Add pre process and post process interface methods.

Signed-off-by: Xinyu Huang <xinyu.huang@intel.com>

### CODE REVIEW

File: tools/crimson/rados_bench_tools.py [hunk 1]
```diff
@@ -109,99 +152,253 @@ def analyse(self):
...
+                    value = int(temp_lis[0].replace(",", ""))
+                    result_dic['branch-misses'] = value
+            line = self.result.readline()
+        self.result.close()
+        return result_dic
+    @staticmethod
+    def post_process(env, test_case_result):
+        # cycle per op
+        result = test_case_result
+        result['cpu_cycle_per_op'] /= (result['client_num'] * result['thread_num'] * int(env.args.time))
```
Comment (cyx1231st) [inline tools/crimson/rados_bench_tools.py:205]: This calculation doesn't make sense to me.

File: tools/crimson/README.rst [hunk 2]
```diff
@@ -79,35 +79,49 @@ Rados bench stress tool
...
 
 Example of result:
 
 .. code-block:: console
 
-   bandwidth                iops             latency reactor_utilization          thread_num          client_num
-   0.5301985               135.5           0.0073715   2.370669279999993                   1                   2
-   0.4831115               123.5          0.01609335  27.892383860000024                   2                   2
+    write_bandwidth             write_iops        write_latency  read_bandwidth           read_iops        read_latency reactor_utilization 
+    15.789866666666668 4041.6666666666665 0.0004911410000000001         17.7848              4552.0         0.000433761   90.82544727999998
```
Comment (cyx1231st) [inline tools/crimson/README.rst:125]: I think it's better to show the total iops/bandwidth to compare the results under different thread/client numbers.

File: tools/crimson/rados_bench_tools.py [hunk 3]
```diff
@@ -109,99 +152,253 @@ def analyse(self):
...
+    def before_run_case(self):
+        self.general_pre_processing()
+        self.pre_processing()
+    def after_run_case(self, test_case_result):
+        self.post_processing(test_case_result)
+        self.general_post_processing()
+    # def before_run(self):
+        # pass
+    # def after_run(self):
+        # pass
```
Comment (cyx1231st) [inline tools/crimson/rados_bench_tools.py:387]: remove unused code.

File: tools/crimson/rados_bench_tools.py [hunk 4]
```diff
@@ -214,30 +411,69 @@ def output(self, output):
...
+            default = False,
+            help = '')
+    parser.add_argument('--seastore',
+            type = bool,
+            default = True,
+            help = '')
+    # parser.add_argument('--restart-env-for-ervery-case',
+            # type = bool,
+            # default = True,
+            # help = '')
```
Comment (cyx1231st) [inline tools/crimson/rados_bench_tools.py:433]: remove unused code.

File: tools/crimson/rados_bench_tools.py (hunk 1 above)
Comment (aisakaki) [reply tools/crimson/rados_bench_tools.py]: I removed this calculation.

File: tools/crimson/README.rst (hunk 2 above)
Comment (aisakaki) [reply tools/crimson/README.rst]: now the outputs will show the total iops/bandwidth in the same type of base case threads test (such as read or write)

File: tools/crimson/README.rst [hunk 5]
```diff
@@ -79,35 +79,58 @@ Rados bench stress tool
...
-utilization of the cpu from seastar. The test thread will run in processors 
-16~31. In consideration of SeaStore starts in processor 0 by default, please 
-avoid setting --taskset to 0.
+    sudo ./rados_bench_tools.py \ 
+        --client-list 4 8 --thread-list 2 4 6 --taskset 16-31 --time 300 \ 
+        --write True --write-ratio 0.75 \
+        --rand-read True --rand-read-ratio 0.25 \ 
+        --reactor-utilization True \
+        --perf True \
+        --crimson True --seastore True
```
Comment (cyx1231st) [inline tools/crimson/README.rst:109]: Can we combine these into meaningful enums? Such as "--scenario" and choose from one of [crimson-seastore, crimson-cyanstore, classic-memstore, classic-bluestore]?

File: tools/crimson/README.rst [hunk 6]
```diff
@@ -79,35 +79,58 @@ Rados bench stress tool
...
+
+The tool will run rados bench write and read test case with the combination 
+of 4 or 8 clients and 2, 4 or 6 threads. In Every test case, the write clients
+will be 75% in all clients and the read clients will be 25%.
+Meanwhile, it will collect the reactor cpu utilization in the case, and the cpu 
+cycle per op of process 2333. The test thread will run in processors 16~31.
```
Comment (cyx1231st) [inline tools/crimson/README.rst:115]: Remove cycle per op.

File: tools/crimson/README.rst [hunk 7]
```diff
@@ -79,35 +79,58 @@ Rados bench stress tool
...
+    read_bandwidth      2.84725               2.68286              3.05852              1.471667
+    read_iops           728.0                 686.0                782.0                376.0
+    read_latency        0.00136141            0.002904295000000000 0.00254781           0.01060875
+    reactor_utilization 51.76855957999997     63.44185818000002    62.81135658000002    57.04136848000001
+    context-switches    21343                 20060                20770                18164
+    cpu_cycle           4617829192            4504157482           4539431704           4732829464
+    instructions        7146125116            7342745221           7228392400           7705910181
+    branches            1305636128            1332912452           1316155264           1385922457
+    thread_num          1                     1                    2                    2
+    client_num          4                     8                    4                    8
```
Comment (cyx1231st) [inline tools/crimson/README.rst:136]: Looks great!

**Review by cyx1231st — APPROVED** [review]
LGTM except some nits.

File: tools/crimson/README.rst (hunk 5 above)
Comment (aisakaki) [reply tools/crimson/README.rst]: OK, I have combined these parameters into one

File: tools/crimson/rados_bench_tools.py [hunk 8]
```diff
@@ -109,99 +162,275 @@ def analyse(self):
...
+    def general_pre_processing(self):
+        # vstart. change the command here if you want to set other start params
+        command = "sudo MGR=1 MON=1 OSD=1 MDS=0 RGW=0 ../src/vstart.sh -n -x \
+                --without-dashboard "
+        scenario = args.scenario
+        if scenario == "crimson-seastore":
+            command += "--crimson --seastore"
+            command += " --seastore-devs " + args.dev
+        elif scenario == "crimson-cyanstore":
+            command += "--crimson --memstore"
```
Comment (cyx1231st) [inline tools/crimson/rados_bench_tools.py:373]: I think its "--crimson --cyanstore"

Comment (aisakaki) [reply tools/crimson/rados_bench_tools.py]: I fixed it

**Comment by aisakaki** [comment]
fixed a small bug in new push.

**Pull request merged**